### DISCUSSION TITLE
Issue installing on Librem5

### DISCUSSION THREAD

**Discussion by programmin1** [ezra-bible-app/ezra-bible-app#644] [Development]
There's an issue installing this on the Librem phone. I tried the debian, Debian11 and Ubuntu .deb files (why are there three??),

none would install and using dpkg -i ezra.deb file I could see the reason, these are build for amd64, and not ARM64.

When can we get some ARM64 builds please?

**Comment by tobias-klein** [comment]
I didn't know about the Librem phone, I first had to google it. How big is the screen of that phone? The app is not made for small screens. Anything below 7" is not suitable with the currently available UI. If you want an ARM build, the first thing that comes to my mind is: Does Electron support that?

Regarding the different Deb files, it is for different versions of Ubuntu.

**Comment by programmin1** [comment]
I suppose I can try the build steps and see... 
You never heard of Libremphone/pinephone? There are also other Linux phones - https://linuxstans.com/linux-phone/ and they are probably going to become more popular considering what's happening with Apple/Google lately - did you see the latest issues with security, that they are not patching on all but the latest phones? https://www.securityweek.com/androids-march-2022-security-updates-patch-39-vulnerabilities

**Comment by tobias-klein** [comment]
I knew that there were some independent Linux based projects, but I had not heard about this particular one.

Give it a try. Contributions are welcome. However, please consider the constraints about the UI on small screens.

**Comment by programmin1** [comment]
I got it to build but now I have this error on "npm start":
```
npm start

> ezra-bible-app@1.6.0-Dev start /home/purism/ezra-bible-app
> electron .

Working with userDataPath /home/purism/.config/ezra-bible-app
Using settings file /home/purism/.config/ezra-bible-app/config.json
[29282:0312/232819.399783:ERROR:web_contents_preferences.cc(261)] preload script must have absolute path.
[29314:0312/232819.974203:ERROR:viz_main_impl.cc(188)] Exiting GPU process due to errors during initialization
[29361:0312/232820.116764:ERROR:sandbox_linux.cc(377)] InitializeSandbox() called with multiple threads in process gpu-process.
Using default database dir /home/purism/.config/ezra-bible-app for database access!
Initializing database at /home/purism/.config/ezra-bible-app
Error occurred in handler for 'initIpc': Error: Please install sqlite3 package manually
    at ConnectionManager._loadDialectModule (/home/purism/ezra-bible-app/node_modules/sequelize/lib/dialects/abstract/connection-manager.js:88:15)
    at new ConnectionManager (/home/purism/ezra-bible-app/node_modules/sequelize/lib/dialects/sqlite/connection-manager.js:22:21)
    at new SqliteDialect (/home/purism/ezra-bible-app/node_modules/sequelize/lib/dialects/sqlite/index.js:14:30)
    at new Sequelize (/home/purism/ezra-bible-app/node_modules/sequelize/lib/sequelize.js:334:20)
    at DbHelper.getSequelize (/home/purism/ezra-bible-app/app/backend/database/db_helper.js:112:21)
    at DbHelper.migrateDatabase (/home/purism/ezra-bible-app/app/backend/database/db_helper.js:117:26)
    at DbHelper.initDatabase (/home/purism/ezra-bible-app/app/backend/database/db_helper.js:40:16)
    at IpcDbHandler.initDatabase (/home/purism/ezra-bible-app/app/backend/ipc/ipc_db_handler.js:52:20)
    at IPC.initDatabase (/home/purism/ezra-bible-app/app/backend/ipc/ipc.js:81:31)
    at IPC.init (/home/purism/ezra-bible-app/app/backend/ipc/ipc.js:74:20)
    at /home/purism/ezra-bible-app/main.js:140:24
```

**Reply by tobias-klein** [reply]
How have you been building the app exactly? On which operating system?

I have seen this issue many times before related to the sqlite3 package, particularly on Windows.
On Linux the build process is usually using pre-built sqlite3 packages, but if you need to compile it for your specific build target then you could do something we also use for the Windows build:

`npm install sqlite3@5.0.0 --build-from-source --runtime=electron --target=17.1.0 --dist-url=https://electronjs.org/headers --target_arch=ia32`

Of course `target_arch` would need to be adjusted.

**Reply by programmin1** [reply]
Arch is aarch64 and I get the same error.

**Reply by tobias-klein** [reply]
Sorry to hear that! What OS are you using? What machine/device are you executing the app on?

**Reply by programmin1** [reply]
I am running those build and run commands on the Purism Librem5. It only gets to the loading screen with that console error.

**Reply by tobias-klein** [reply]
The error message indicates that the sqlite module is not binary-compatible with the Electron binary. If the above mentioned build command did not help then the only other option is to further explore the build process of sqlite3 and potentially Electron and maybe ask the [node-sqlite3](https://github.com/mapbox/node-sqlite3) folks for help.

**Reply by programmin1** [reply]
It runs and gets past loading... screen after "npm update". :)

**Reply by tobias-klein** [reply]
Cool! 👍 
Which exact node and npm version are you using now?

**Reply by programmin1** [reply]
Please see https://github.com/ezra-bible-app/ezra-bible-app/pull/646 I also had to make some adjustments due to the small screen.

**Reply by tobias-klein** [reply]
So, when you ran npm update, several packages got updated, including sqlite3. I am guessing that the sqlite3 update was the critical one. Can you revert all the other updates and try whether that's sufficient?

**Comment by programmin1** [comment]
This was working for Librem phone so I wonder if you could try adding a debian build with --platform linux/arm64

it "should" build an arm64 from what I read here https://stackoverflow.com/questions/47809904/how-to-set-architecture-for-docker-build-to-arm64

I am not sure how I could run this manually.

**Comment by sammyj85** [comment]
Pinephone user here - I'm glad I'm not alone in pushing for a Bible app for Linux ARM64. I made do with a desktop app in the meantime...

Manually building worked first time, and the app looks great on a 6 inch screen (150% scaling) from my initial play around! Thanks!

ARM64 is emerging as a more popular hardware platform, so it might be helpful to name the release packages as x86_64. I downloaded the one matching my phone's OS without thinking too much, but of course it gave an error that it was not arm64.

**Comment by tobias-klein** [comment]
@sammyj85 
Thanks for your comments! Could you imagine to contribute an ARM64 package?

What's needed for that is a Docker image that contains the build environment (see [here](https://github.com/ezra-bible-app/ezra-bible-app-docker/tree/master/docker)) as well as adaptions in the build scripts/GitHub actions configuration.

I'd be glad to see any pull requests!

Regarding the layout: Various adaptions were made in the mobile layout (for smaller screen resolutions) recently. Those optimizations should work/apply on all platforms, so even if you use the Desktop app on your phone - if it is scaled to your screen it should still (mostly) behave like the app on Android mobile.

**Reply by sammyj85** [reply]
Thanks for your reply and your openness. I'll look into it. I've done cross compiles for arm64 from x86_64 before, but I wouldn't imagine cross-packaging would be easy. I wasn't previously sure what kinds of platforms were available for github CI, but it may well be.

It looks like the mobile usage optimisations have worked a treat. I realise a lot of time has passed since the original comment, so it's great to see it progress well. The desktop reference earlier was for a different program, BibleTime (definitely desktop oriented).

**Reply by tobias-klein** [reply]
Once you have the Docker image, the rest should be quite straightforward.

I could write a short howto for adding new packages in general, outlining all the steps. This could be added to the [contribution guidelines](https://github.com/ezra-bible-app/ezra-bible-app/blob/master/CONTRIBUTING.md).
